Mathlib/Combinatorics/Additive/CovBySMul.lean

```diff
@@ -0,0 +1,70 @@
+/-
+Copyright (c) 2024 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.Group.Pointwise.Finset.Basic
+import Mathlib.Data.Real.Basic
+
+/-!
+# Relation of covering by cosets
+
+This file defines a predicate for a set to be covered by at most `K` cosets of another set.
+
+This is a fundamental relation to study in additive combinatorics.
+-/
+
+open scoped Finset Pointwise
+
+variable {M N X : Type*} [Monoid M] [Monoid N] [MulAction M X] [MulAction N X] {K L : ℝ}
+  {A A₁ A₂ B B₁ B₂ C : Set X}
+
+variable (M) in
+/-- Predicate for a set `A` to be covered by at most `K` cosets of another set `B` under the action
+by the monoid `M`. -/
+@[to_additive "Predicate for a set `A` to be covered by at most `K` cosets of another set `B` under
+the action by the monoid `M`."]
+def CovBySMul (K : ℝ) (A B : Set X) : Prop := ∃ F : Finset M, #F ≤ K ∧ A ⊆ (F : Set M) • B
```

Comment:
There remains the question of if this should be a Nat instead...

Reply:
Here I would say "Revisit later?". GrowthInGroups is currently using `CovBySMul` as a convenience tool but hasn't yet used it in a crucial way that shows `ℝ` or `ℕ` is particularly better. My current use of `ℝ` is motivated by the literature setting `K : ℝ` when talking about approximate subgroups, and I therefore have a slight bias towards thinking this is better.

Reply:
Yup, that's fine with me.